src/librustc_resolve/lib.rs

```diff
@@ -2644,14 +2608,38 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         ResolveAttempt(resolution)
     }
 
-    /// Skips `path_depth` trailing segments, which is also reflected in the
-    /// returned value. See `middle::def::PathResolution` for more info.
     pub fn resolve_path(&mut self,
                         id: NodeId,
                         path: &Path,
                         path_depth: usize,
                         namespace: Namespace)
                         -> Option<PathResolution> {
+        let resolution = self.resolve_path_noprim(id, path, path_depth, namespace);
+        // Check if the resolution is a #[primitive_type] item in type namespace,
+        // replace that item with a primitive type with the same name.
+        if let Some(PathResolution{base_def, ..}) = resolution {
+            if let Some(def_id) = base_def.opt_def_id() {
```

Comment:
I'd prefer to check for `Def::Err` explicitly instead of introducing the method `opt_def_id`.